staging/src/k8s.io/client-go/tools/portforward/tunneling_dialer.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package portforward
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/websocket"
+	"k8s.io/klog/v2"
+)
+
+// tunnelingDialer implements "httpstream.Dial" interface
+type tunnelingDialer struct {
+	url       *url.URL
+	transport http.RoundTripper
+	holder    websocket.ConnectionHolder
+	ports     []string
+}
+
+// NewTunnelingDialer creates and returns the tunnelingDialer structure which implemements the "httpstream.Dialer"
+// interface. The dialer can upgrade a websocket request, creating a websocket connection. This function
+// returns an error if one occurs.
+func NewTunnelingDialer(url *url.URL, config *restclient.Config, ports []string) (httpstream.Dialer, error) {
+	if len(ports) == 0 {
+		return nil, fmt.Errorf("you must specify at least 1 port")
+	}
+	transport, holder, err := websocket.RoundTripperFor(config)
+	if err != nil {
+		return nil, err
+	}
+	return &tunnelingDialer{
+		url:       url,
+		transport: transport,
+		holder:    holder,
+		ports:     ports,
+	}, nil
+}
+
+// Dial upgrades a websocket request, returning a websocket connection (wrapped
+// by an "httpstream.Connection"), the negotiated protocol, or an error if one occurred.
+func (d *tunnelingDialer) Dial(protocols ...string) (httpstream.Connection, string, error) {
+	// There is no passed context, so skip the context when creating request for now.
+	// Websockets requires "GET" method: RFC 6455 Sec. 4.1 (page 17).
+	req, err := http.NewRequest("GET", d.url.String(), nil)
+	if err != nil {
+		return nil, "", err
+	}
+	// Add the port(s) as request query params.
+	forwardedPorts, err := parsePorts(d.ports)
+	if err != nil {
+		return nil, "", err
+	}
+	for _, port := range forwardedPorts {
+		query := req.URL.Query()
+		remotePort := int(port.Remote)
+		klog.V(4).Infof("Remote Port: %d", remotePort)
+		query.Set("ports", strconv.Itoa(remotePort))
+		req.URL.RawQuery = query.Encode()
+	}
+	// Hard-code the v2 portforward protocol for the websocket dialer for now.
+	websocketProtocols := []string{"v2.portforward.k8s.io"}
```

Comment:
the protocols passed in are spdy protocols... this dialer is going to tunnel those over websockets, so I think we could construct the websocket protocol name generically by prefixing `spdy+`, something like this:

```
var spdyTunnelingWebsocketProtocols []string
for _, p := range protocols {
  spdyTunnelingWebsocketProtocols = append(spdyTunnelingWebsocketProtocols, "SPDY/3.1+"+p)
}
conn, err := websocket.Negotiate(d.transport, d.holder, req, websocketProtocols...)
...
```

then whatever the server sends back as the accepted websocket subprotocol, we trim off the `"SPDY/3.1+"` prefix for, and then can return the chosen spdy protocol in a way that makes sense to the caller
```
selectedSpdyProtocol := strings.TrimPrefix(conn.Subprotocol(), "SPDY/3.1+")
...
return spdyConn, selectedSpdyProtocol, err
```

that lets this client tunneler be completely agnostic about whatever SPDY/3.1 stuff is happening over the tunnel

Reply:
We're adding the SPDY protocols (and other request headers) as we call the `UpgradeAwareProxy` passing in the updated `http.Request`. The `UpgradeAwareProxy` is doing the SPDY upgrade using these modified headers. We don't need to do anything special with these WebSocket upgrade headers.